### PR TITLE
[embedded] Use `__has_feature(swiftcc)` to detect Swift calling convention

### LINKED ISSUE #72343: Embedded `wasm32-unknown-none-wasm` incorrectly calls class destructor with C_CC
### Description

We check availability of `swiftasynccc` feature to detect if the target support `swiftcc` here https://github.com/apple/swift/blob/735e28b551e13262c2ff0ce013608a2202d5b8b6/stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h#L32

But wasm target doesn't support `swiftasynccc` but does `swiftcc`, so `_swift_embedded_invoke_heap_object_destroy` calls destructor functions without swiftcc even though compiler emits destructor definitions with swiftcc.

### Reproduction

```swift
// RUN: split-file %s %t
// RUN: %target-swiftc_driver %t/check.swift -Xclang-linker %t/rt.c -target wasm32-unknown-none-wasm \
// RUN:   -enable-experimental-feature Embedded -Xcc -fdeclspec -wmo -Xfrontend -disable-stack-protector \
// RUN:   -Xclang-linker -nostdlib -o %t/check.wasm
// RUN: %target-run %t/check.wasm

// -- rt.c

#include <stddef.h>
#include <stdint.h>

void free(void *ptr) {}

int posix_memalign(void **memptr, size_t alignment, size_t size) {
    uintptr_t mem = __builtin_wasm_memory_grow(0, (size + 0xffff) / 0x10000);
    if (mem == -1) {
        return -1;
    }
    *memptr = (void *)(mem * 0x10000);
    *memptr = (void *)(((uintptr_t)*memptr + alignment - 1) & -alignment);
    return 0;
}

// -- check.swift

class Foo {}

@_cdecl("_start")
func main() {
  _ = Foo()
}
```

### Expected behavior

No CC mismatch

### Environment

3d4457b63226e3330f893f1af985965cc80839aa

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by kateinoigakukun**
This change replaces the use of `__has_feature(swiftasynccc)` with `__has_feature(swiftcc)` to detect the SwiftCC availability. The former condition works fine for most platforms that support both SwiftCC and SwiftAsyncCC or neither, but it fails for WebAssembly, which supports SwiftCC but not SwiftAsyncCC.

This change depends on https://github.com/llvm/llvm-project/pull/85347

Close https://github.com/apple/swift/issues/72343

### CODE REVIEW

**Comment by kubamracek** [comment]
If https://github.com/llvm/llvm-project/pull/85347/files causes `__has_feature(swiftasynccc)` to stop working, then we also need to adjust Runtime/Config.h.

**Comment by kateinoigakukun** [comment]
@swift-ci Please test with the following PR
https://github.com/apple/llvm-project/pull/8441

**Comment by kateinoigakukun** [comment]
@swift-ci Please test WebAssembly
https://github.com/apple/llvm-project/pull/8441

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

**Review by kubamracek — APPROVED** [review]

File: test/embedded/classes-wasm.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// RUN: %empty-directory(%t)
+// RUN: split-file %s %t
+// RUN: %swift-frontend -c %t/check.swift -parse-as-library -target wasm32-unknown-none-wasm \
+// RUN:   -enable-experimental-feature Embedded -Xcc -fdeclspec -disable-stack-protector \
+// RUN:   -o %t/check.o
+// RUN: %clang -target wasm32-unknown-none-wasm %t/check.o %t/rt.c -nostdlib -o %t/check.wasm
+// RUN: %target-run %t/check.wasm
+// REQUIRES: executable_test
+// REQUIRES: CPU=wasm32
```
Comment (kubamracek) [inline test/embedded/classes-wasm.swift:9]: Can you point me to a CI job that actually runs these executable WASM tests?

File: stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h [hunk 2]
```diff
@@ -28,8 +28,7 @@
 extern "C" {
 #endif
 
-// FIXME: Replace with __has_feature(swiftcc) once that's added to Clang.
-#if __has_feature(swiftasynccc)
+#if __has_extension(swiftcc)
```
Comment (aschwaighofer) [inline stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h:31]: Should we use `__has_feature(swiftasynccc) || __has_extension(swiftcc)` to be compatible with older compilers for a while?

@compnerd Do all windows builds use a just built compiler such that this is not a concern?

File: test/embedded/classes-wasm.swift (hunk 1 above)
Comment (kateinoigakukun) [reply test/embedded/classes-wasm.swift:9]: Wasm CI runs executable tests on WasmKit runtime: https://ci.swift.org/job/swift-PR-Linux-crosscompile-wasm/31

File: stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h (hunk 2 above)
Comment (kateinoigakukun) [reply stdlib/public/SwiftShims/swift/shims/EmbeddedShims.h:31]: I think it's worth being conservative here.

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

**Comment by kateinoigakukun** [comment]
Hmm, `#if __has_attribute(swiftasynccall) && (__has_feature(swiftasynccc) || __has_extension(swiftasynccc))` is not acceptable with MSVC cl.exe?

https://ci-external.swift.org/job/swift-PR-windows/23188/console
```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\include\swift/Runtime/Config.h(231): fatal error C1012: unmatched parenthesis: missing ')'
```

```c
  // SWIFT_CC(swiftasync) is the Swift async calling convention.
  // We assume that it supports mandatory tail call elimination.
  #if __has_attribute(swiftasynccall) && (__has_feature(swiftasynccc) || __has_extension(swiftasynccc))
  #define SWIFT_CC_swiftasync __attribute__((swiftasynccall))
  #else
  #define SWIFT_CC_swiftasync SWIFT_CC_swift
  #endif
```

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

**Comment by kateinoigakukun** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/8441

@swift-ci Please test WebAssembly

**Pull request merged**